### PR TITLE
Gameobject rotation (from mangos/MaNGOS@2bcbc0f)

### PR DESCRIPTION
**Pull request by ariel-**
Ported functions and merged with newer core.

### CODE REVIEW

**Comment by Goatform** [comment]
nice :+1:

**Comment by Naios** [comment]
:+1:

**Comment by Aokromes** [comment]
Anyone can test it and move around world searching for incorrectly spawned doors or other things?
Imho better fix db before push this taking data from UDB.

**Comment by NorthStrider** [comment]
currently compiling and testing. gonna give feedback asap im done with some tests

**Comment by NorthStrider** [comment]
well it appears that all doors are facing into 1 direction:
i rechecked the sniffs where the db spawns for shadow fang keep were comming from but it seems sth is missing (tdb spawns ofc, sfk was spawned by me back in closed tdb 434 times)

wowscrnshot_021415_185301
wowscrnshot_021415_184131
wowscrnshot_021415_184151

**Comment by ariel-** [comment]
Many GOs are fine, some slanted examples here:

wowscrnshot_021515_011054
wowscrnshot_021515_011258
wowscrnshot_021515_011338
wowscrnshot_021515_011440
wowscrnshot_021515_011453

wowscrnshot_021515_011702

**Comment by Golrag** [comment]
Maybe the SQL in the comment from the mangos commit can help
https://github.com/mangos/MaNGOS/commit/2bcbc0fbfe63fab389e4f4f875ac215852bff8b6
```sql
UPDATE gameobject SET rotation0 = 0, rotation1 = 0, rotation2 = SIN(orientation * 0.5), rotation3 = COS(orientation * 0.5);
```
Not sure if this is Blizzlike :/

**Comment by Aokromes** [comment]
We have a problem, we have at least some GOs with proper rotation, we need to compare go id + x + y + z from TDB GOs to UDB GOs.

File: src/server/scripts/Commands/cs_gobject.cpp [hunk 1]
```diff
@@ -416,25 +416,37 @@ class gobject_commandscript : public CommandScript
...
-        object->UpdateRotationFields();
+        // try extract two opt-parameters
+        orientation = strtok(NULL, " ");
+        if (orientation)
+        {
+            oy = (float)atof(orientation);
+            if (orientation = strtok(NULL, " "))
+                ox = (float)atof(orientation);
+            else
+                ox = 0;
```
Comment (jackpoz) [inline src/server/scripts/Commands/cs_gobject.cpp]: please use 0.f to clearly state the type of the number

**Comment by CreshandKesh** [comment]
Fixes Dalaran, Data Taken from UDB
```sql
UPDATE gameobject SET rotation0=-0.010067,  rotation1=-0.0192642, rotation2=-0.54460,  rotation3=0.838411  WHERE guid=57794;
UPDATE gameobject SET rotation0=-0.0478525, rotation1=-0.0228243, rotation2=-0.90131,  rotation3=0.429906  WHERE guid=57849;
UPDATE gameobject SET rotation2=0.143493,   rotation3=0.989651  WHERE guid=60113;
UPDATE gameobject SET rotation2=0.25038,    rotation3=0.968148  WHERE guid=60120;
UPDATE gameobject SET rotation2=0.309016,   rotation3=0.951057  WHERE guid=60227;
UPDATE gameobject SET rotation2=0.426569,   rotation3=0.904455  WHERE guid=60244;
UPDATE gameobject SET rotation2=-0.902585,  rotation3=0.430512  WHERE guid=60253;
UPDATE gameobject SET rotation2=0.426569,   rotation3=0.904455  WHERE guid=60258;
UPDATE gameobject SET rotation2=0.945518,   rotation3=0.325568  WHERE guid=60297;
UPDATE gameobject SET rotation2=0.434444,   rotation3=0.900699  WHERE guid=60323;
UPDATE gameobject SET rotation2=0.450098,   rotation3=0.892979  WHERE guid=61476;
UPDATE gameobject SET rotation2=0.588305,   rotation3=0.808639  WHERE guid=56109;
UPDATE gameobject SET rotation2=-0.90258,   rotation3=0.430512  WHERE guid=57802;
UPDATE gameobject SET rotation2=-0.78801,   rotation3=0.615662  WHERE guid=57892;
UPDATE gameobject SET rotation2=0.999391,   rotation3=0.0349061 WHERE guid=57913;
UPDATE gameobject SET rotation2=-0.63607,   rotation3=0.771625  WHERE guid=57960;
UPDATE gameobject SET rotation2=0.999391,   rotation3=0.0349061 WHERE guid=58082;
UPDATE gameobject SET rotation2=0.130525,   rotation3=0.991445  WHERE guid=60303;
UPDATE gameobject SET rotation2=-0.808639,  rotation3=0.588306  WHERE guid=56101;
UPDATE gameobject SET rotation2=-0.824125,  rotation3=0.566408  WHERE guid=61486;
UPDATE gameobject SET rotation2=0.566406,   rotation3=0.824126  WHERE guid=61497;
UPDATE gameobject SET rotation2=0.430511,   rotation3=0.902586  WHERE guid=61606;
```

File: src/server/scripts/Commands/cs_gobject.cpp [hunk 2]
```diff
@@ -416,25 +416,37 @@ class gobject_commandscript : public CommandScript
...
             Player* player = handler->GetSession()->GetPlayer();
-            o = player->GetOrientation();
+            oz = player->GetOrientation();
         }
 
-        object->Relocate(object->GetPositionX(), object->GetPositionY(), object->GetPositionZ(), o);
-        object->UpdateRotationFields();
+        // try extract two opt-parameters
+        orientation = strtok(NULL, " ");
+        if (orientation)
```
Comment (jackpoz) [inline src/server/scripts/Commands/cs_gobject.cpp]: https://travis-ci.org/TrinityCore/TrinityCore/builds/50880616#L1045
>variable 'ox' is used uninitialized whenever 'if' condition is false

was this part tested ingame (or even debugged) ?

**Comment by Aokromes** [comment]
NorthStrider / Ariel- can you check dalaran with this patch after apply https://github.com/TrinityCore/TrinityCore/pull/14146#issuecomment-74441429 ?

**Comment by ariel-** [comment]
Did a before and after run and it fixes those guids, guild banks are still circa 90 degrees into the wall tho

**Comment by CreshandKesh** [comment]
Now I see, also missing several objects such as chairs and mailboxes, working on it. thanx @ariel-

**Comment by CreshandKesh** [comment]
Fixes :
https://gist.github.com/CreshandKesh/c3dcf0f8106ac84ce82f

**Comment by Aokromes** [comment]
IMHO the higher priority is fix doors on instances/entrances to instances before pushing this.

**Comment by Aokromes** [comment]
```sql
select * from udb.gameobject as t where t.id in (select id from world.gameobject as u inner join world.gameobject_template as ugt on u.id = ugt.entry where ugt.type = 0 and u.id = t.id and u.position_x = t.position_x and  u.position_y = t.position_y and  u.position_z = t.position_z and (u.rotation0 != t.rotation0 or u.rotation1 != t.rotation1 or u.rotation2 != t.rotation2 or u.rotation3 != t.rotation3));

619 rows in set (15.98 sec)

**Comment by Golrag** [comment]
What about this commit ?https://github.com/mangos/MaNGOS/commit/0fe88f35dfb2e4f9454d160cdad439e7acac326e

**Comment by Naios** [comment]
@ariel- probably you should take a look at the current head of mangos gameobject rotation to include their whole work.

**Comment by ariel-** [comment]
True that, probably trying to port HEAD is a better idea, I'll keep updating this

**Comment by ariel-** [comment]
Transports issue: MaNGOS maintained a table in DB with custom rotation values for transports, as of now, applying this wrecks the deeprun tram eg (goes the other way around)

File: src/server/game/Entities/GameObject/GameObject.h [hunk 3]
```diff
@@ -32,6 +32,18 @@ class Transport;
 
 #define MAX_GAMEOBJECT_QUEST_ITEMS 6
 
+// small wrapper to avoid clutter
+struct QuatData
```
Comment (jackpoz) [inline src/server/game/Entities/GameObject/GameObject.h]: would it make sense to use G3D Quaternion ?

**Comment by ariel-** [comment]
Idk, that's how it was on mangos, maybe to avoid extra overhead? since they don't need all the extra operations/members a full fledged G3D quat has to offer, they only need the 4 coords.

**Comment by jackpoz** [comment]
G3D::Quat is used at https://github.com/TrinityCore/TrinityCore/pull/14146/files#diff-fe5c8ebb7ad8e663836b00f39eb524d6R1856 and at https://github.com/TrinityCore/TrinityCore/pull/14146/files#diff-fe5c8ebb7ad8e663836b00f39eb524d6R1879 tho, it would be nice to be consistent all around the code

**Comment by Arsul** [comment]
Hope this helps

Copied udb.gameobject and tdb.gameobject, then used the following query, which checks for same identity (guid, id, x, y, z) but different rotation values. Should I include Orientation into the Join too?


select t.guid, t.id, t.map, t.spawnmask, t.phasemask, t.position_x, t.position_y, t.position_z, t.orientation, u.rotation0, u.rotation1, u.rotation2, u.rotation3, t.spawntimesecs, t.animprogress, t.state from gameobject as t 
JOIN gameobject.gameobjectcmangos as u on u.guid=t.guid and u.id=t.id and u.position_x=t.position_x and u.position_y=t.position_y and u.position_z=t.position_z
WHERE ((u.rotation3 - t.rotation3 > 0.01) OR (u.rotation2 - t.rotation2 > 0.01) OR (u.rotation1 - t.rotation1 > 0.01) OR (u.rotation0 - t.rotation0 > 0.01) OR (u.rotation3 - t.rotation3 < -0.01) OR (u.rotation2 - t.rotation2 < -0.01) OR (u.rotation1 - t.rotation1 < -0.01) OR (u.rotation0 - t.rotation0 < -0.01))

Output where 47637 Lines with different values. Some Copy&Paste Magic and the following SQL File was the result:

http://www.file-upload.net/download-10515873/Gameobjectrotation.sql.html

Any suggestions?

File: src/server/game/Entities/GameObject/GameObject.h (hunk 3 above)
Comment (armano2) [inline src/server/game/Entities/GameObject/GameObject.h]: yes G3D::Quat should be used here

**Comment by DDuarte** [comment]
Deleting and inserting a ton of gameobjects where a update would be enough might not be a great idea.

http://software.dell.com/products/toad-for-mysql/ can be used to compare two DBs and generate UPDATE queries.

**Comment by ariel-** [comment]
Maybe something like [this](https://github.com/TrinityCore/TrinityCore/issues/13750#issuecomment-73254334) can be used (I create a temporary table with all the updates, populate it only with the columns that are changed, then update and drop)

**Comment by jackpoz** [comment]
any progress on the sql part of this pr ?

**Comment by ariel-** [comment]
Is it ok to automatically asume that UDB has true rotation values and TDB does not? or will I need some sniffs?

**Comment by Nyeriah** [comment]
It has been a decade but doesn't UDB have VerifiedBuild too? You could check that, though checking sniffs would actually be better

**Comment by ariel-** [comment]
Updated, still missing some GO's (specially those that were nearby but not exactly on the same position in each DB)

**Comment by Aokromes** [comment]
@ariel- :) and jump into irc xd
This branch has conflicts that must be resolved
 Use the command line to resolve conflicts before continuing.

**Comment by tkrokli** [comment]
The PR branch `ariel-:gobrot` is 520 commits behind `TrinityCore:3.3.5` - needs to be updated.

**Comment by msoky** [comment]
can be merged?

**Comment by Rushor** [comment]
Updated, still missing some GO's (specially those that were nearby but not exactly on the same position in each DB)

this needs to be done first, else we will have thousands of userreports concerning wrong go positions

**Comment by jackpoz** [comment]
I heard this change should have a PR on WPP too otherwise it will be a mess with 2 different structures

**Comment by Shauren** [comment]
Instead of making this large sql file, why didn't you just write 3 queries like this
```
UPDATE gameobject_addon ga JOIN gameobject g ON ga.guid = g.guid SET ga.path_rotation0 = g.rotation0, ga.path_rotation1 = g.rotation1, ga.path_rotation2 = g.rotation2, ga.path_rotation3 = g.rotation3;
INSERT IGNORE INTO gameobject_addon SELECT guid, rotation0, rotation1, rotation2, rotation3, 0, 0 FROM gameobject WHERE rotation0 != 0 OR rotation1 != 0 OR rotation2 != 0 OR rotation3 != 1;
UPDATE gameobject SET rotation0 = 0, rotation1 = 0; -- existing core calculation always ended up with this
--- 3rd query calculates unpacked rotation2 and rotation3 from orientation like this https://github.com/TrinityCore/TrinityCore/blob/6.x/src/server/game/Entities/GameObject/GameObject.cpp#L1958 ---
```
And then finally after that, port any mangos db rows that have rotation0 != 0 or rotation1 != 0 - there are only 565 of them in your sql, compared to the current thousands

**Comment by ariel-** [comment]
I don't entirely like it, it encourages for bad future spawning of GOs with this code in place.

https://github.com/TrinityCore/TrinityCore/pull/14146/files#diff-fe5c8ebb7ad8e663836b00f39eb524d6R221

PARENT_ROTATION field is only for transport type GameObject; it shouldn't be used to hack and send to client existing bad data.

**Comment by Shauren** [comment]
What I meant is porting that method into the one-time sql for existing spawns to minimize potential errors (and obviously remove it from code)

**Comment by ariel-** [comment]
I'm updating this. However I've run into a problem.

You see, rotation fields and GAMEOBJECT_PARENTROTATION are two different things, in sniffs only some gameobjects have GAMEOBJECT_PARENTROTATION sent but all of them have rotations.

The above described behaviour makes it similar to creature_addon that sets fields for some (but not all) spawns, which is exactly what gameobject_addon is for.

There are few 'elegant' options besides re-spawning the gameobject DB, I tried with yours but it'll just ruin the few 'good' spawns TC has (sending them to addon is not an option, or would require ANOTHER set of 4 fields, because as stated before GAMEOBJECT_PARENTROTATION != packed rotation)

So, here comes the "intermediate" approach: Spawn whitelisting
Add a flag in gameobject_addon table: with this flag set, gameobject will use rotation fields from db, if not set or 0, gameobject will be sent to client with current (or legacy) rotation values (ie calculated from orientation)

**Comment by Shauren** [comment]
I still have a feeling you are not understanding what I mean
* Moving current gameobject.rotation fields into gameobject_addon with no alteration is safe - old code loads gameobject.rotation into PARENTROTATION, new code loads gameobject_addon into PARENTROTATION
* Old code has no source for storing packed rotation, instead it calculates it from orientation using this code
```
    static double const atan_pow = atan(pow(2.0f, -20.0f));

    double f_rot1 = std::sin(GetOrientation() / 2.0f);
    double f_rot2 = std::cos(GetOrientation() / 2.0f);

    int64 i_rot1 = int64(f_rot1 / atan_pow *(f_rot2 >= 0 ? 1.0f : -1.0f));
    int64 rotation = (i_rot1 << 43 >> 43) & 0x00000000001FFFFF;

    m_rotation = rotation;
```
* New code will load packed rotation from gameobject.rotation fields - and this is what we need to fill - using the method above (adapted for sql) we calculate gameobject.rotation from orientation
Having this done, all objects should have 100% the same values sent to clients as before

After this is done, the db should have all gameobject.rotation0 and gameobject.rotation1 fields set to 0 because there was no support for rotating around X and Y axes before (spawns that do need it can be ported from mangos db)

**Comment by Warpten** [comment]
I think it would be best to just store orientation as a quat directly in db; after all even floats for rotations don't mean anything to regular humans and you basically save the hassle of dealing with arbitrary precision

**Comment by ariel-** [comment]
What would be a good mysql data type for storing quats directly then?

**Comment by Shauren** [comment]
4x float, ignore warptens random math stuff

**Comment by ariel-** [comment]
That should do it, trouble spots I checked are now gone.

File: src/server/game/Entities/GameObject/GameObject.cpp [hunk 4]
```diff
@@ -1899,34 +1893,46 @@ std::string const & GameObject::GetNameForLocaleIdx(LocaleConstant loc_idx) cons
...
+}
 
-    //float f_rot3 = std::sin(0.0f / 2.0f);
-    //int64 i_rot3 = f_rot3 / atan(pow(2.0f, -21.0f));
-    //rotation |= (i_rot3 >> 42) & 0x7FFFFC0000000000;
+void GameObject::SetWorldRotation(G3D::Quat const& rot)
+{
+    G3D::Quat rotation;
+    // Temporary solution for gameobjects that have no rotation data in DB:
+    if (G3D::fuzzyEq(rot.z, 0.f) && G3D::fuzzyEq(rot.w, 0.f))
```
Comment (Shauren) [inline src/server/game/Entities/GameObject/GameObject.cpp]: This is not needed thanks to the query doing the calculation

**Pull request closed without merging**